### ISSUE TITLE
Can we have save as Scaled FP8 that will work with ComfyUI? for Qwen and Wan trainings - at least for fine tuning

### ISSUE THREAD

**Issue by FurkanGozukara** [kohya-ss/musubi-tuner#634]
Can we have save as Scaled FP8 that will work with ComfyUI? for Qwen and Wan trainings - at least for fine tuning @kohya-ss 
that would be really cool

your Scaled FP8 system is amazing

I added to my Ovi and it works perfect

Scaled with your system, saved and loading from this

<img width="1086" height="294" alt="Image" src="https://github.com/user-attachments/assets/40a98106-8cb5-4510-a609-9abcb93fbe61" />

<img width="1334" height="358" alt="Image" src="https://github.com/user-attachments/assets/931b5aea-e7d7-4add-8126-ecdecca5522f" />

**Comment by Sarania** [comment]
Firstly it's awesome to see that you've taken to implementing stuff yourself, I agree that the fp8_scaled code is awesome and easy to work with! 

So the thing with saving them out like that is if we later want to merge LoRA it presents accuracy problems because we must take the fp8 scaled model and dequantize it to fp16/bf16, merge any LoRA in, and then requantize it to fp8_scaled. If we keep the weights as bf16/fp16 and quantize at runtime then we can merge the LoRA straight away and then quantize only once, preserving maximum quality.

Under the hood, there's also the fact that we have to patch the model's nn.Linear forward passes to hook in our dequantization routine(apply_fp8_monkey_patch). This is especially important now with the blockwise quantization as the weight needs to be reshaped for the computations during forward. So these are the reasons saving them doesn't make much sense.

That said, if you just want to play with fp8_scaled in Comfy the support is there you can find presaved models like [these](https://huggingface.co/Kijai/WanVideo_comfy_fp8_scaled/tree/main) and I'm sure for Qwen as well. I actually tried to implement a node to do musubi style on the fly fp8_scaled in Comfy a couple of months ago. I got it to work but only without LoRA. Comfy patches things in a very specific way that makes stuff like that tricky and I couldn't figure it out >_<

**Comment by FurkanGozukara** [comment]
@Sarania it is true i added LoRA support as well to ovi

I had to use bf16 version of model, apply LoRA and then scale

people just complaining me that saved file sizes are too big :D 

maybe we could save as GGUF?

**Comment by polym** [comment]
@FurkanGozukara I just write a ComfyUI custom node to load musubi fp8_scaled qwen model. Would you like to try it?
https://github.com/polym/ComfyUI-musubi

**Comment by FurkanGozukara** [comment]
> [@FurkanGozukara](https://github.com/FurkanGozukara) I just write a ComfyUI custom node to load musubi fp8_scaled qwen model. Would you like to try it? https://github.com/polym/ComfyUI-musubi

thanks

yesterday generated Qwen Image Edit FP8 Scaled model and it works with native comfyui already :D 

i am the first one probably here my model : https://huggingface.co/MonsterMMORPG/Wan_GGUF/blob/main/Qwen_Image_Edit_Plus_2509_FP8_Scaled.safetensors

so @kohya-ss perhaps we can save like this

i tested and it works with LoRAs perfectly fine and same quality as GGUF Q8 and 1.5x faster inference

**Comment by kohya-ss** [comment]
It is possible to save and load block-wise scaled weights. However, since quantization methods are constantly evolving, I think it would be preferable to keep the weight files in bf16 and quantize them dynamically. Loading speed is also not that much slower.

However, I can understand the advantage of saving storage.
While weights may be deprecated in the future, it might be an idea to be able to save and load weights using the current quantization method.

**Comment by FurkanGozukara** [comment]
> It is possible to save and load block-wise scaled weights. However, since quantization methods are constantly evolving, I think it would be preferable to keep the weight files in bf16 and quantize them dynamically. Loading speed is also not that much slower.
> 
> However, I can understand the advantage of saving storage. While weights may be deprecated in the future, it might be an idea to be able to save and load weights using the current quantization method.

yes would be nice addition and as you said we may quantize better way later

your FP8 Scaling is amazing literally

**Comment by polym** [comment]
> i am the first one probably here my model : https://huggingface.co/MonsterMMORPG/Wan_GGUF/blob/main/Qwen_Image_Edit_Plus_2509_FP8_Scaled.safetensors

@FurkanGozukara Good job!

One more question, It looks like tensor-wise fp8_scaled not block-wise, right?

**Comment by FurkanGozukara** [comment]
> > i am the first one probably here my model : https://huggingface.co/MonsterMMORPG/Wan_GGUF/blob/main/Qwen_Image_Edit_Plus_2509_FP8_Scaled.safetensors
> 
> [@FurkanGozukara](https://github.com/FurkanGozukara) Good job!
> 
> One more question, It looks like tensor-wise fp8_scaled not block-wise, right?

For Qwen Per-tensor quantization
For Wan 2.2 Per-channel quantization

<img width="936" height="201" alt="Image" src="https://github.com/user-attachments/assets/d744634c-bc14-4a2f-bdd5-696b6e2835c2" />

**Comment by Sarania** [comment]
I'm glad to see you all taking an interest in the FP8 scaled implementation too, I've gushed about it many times before. It's compact, well written, sanely commented, and highly portable among models. Also, unlike other notable quantization methods such as TorchAO, I find Kohya's fp8 code to be a lot easier to play with and extend. Also TorchAO can be finnicky but Kohya's FP8 "Just works" anywhere I need it to. Simple options for target/exclude make it easy to optimize what parts of the model you wanna quantize versus don't so you can squeeze every last drop of quality out of whatever VRAM you've got. I love it and I make no secret about it. It was the first quantization code I looked and and went "Oh, this actually makes sense to me!"

Honestly I could envision making a little standalone library of it, where you can import it and feed it a model and get back a ready to go fp8 scaled model. Doing that for just the base models would be easier, supporting LoRA could be trickier but doable. I might consider it at some point, I've been on a lil coding break lately but we'll be back soon enough XD I think it has a lot of value.

**Comment by FurkanGozukara** [comment]
@Sarania ye supporting lora was so hard for me

the solution i found was merge weights on bf16 and then quantize

i don't know how comfyui handling but it is working fine there :D

**Comment by polym** [comment]
<img width="388" height="183" alt="Image" src="https://github.com/user-attachments/assets/5afd50b0-2abf-4015-bb24-4d8c551f45f1" />

After migrated amazing code in fp8_optimization.py to my repo, I added a new comfyui node MusubiUNETJitLoader. Now, we can load qwen_image_bf16 checkpoint, patch lora to it and then fp8 scaling on the fly.

[musubi-lora-example-02.json](https://github.com/user-attachments/files/22966085/musubi-lora-example-02.json)
